v-next/hardhat/src/internal/builtin-plugins/solidity/build-system/compilation-job.ts

```diff
@@ -124,12 +124,31 @@ export class CompilationJobImplementation implements CompilationJob {
         evmVersion:
           settings.evmVersion ??
           getEvmVersionFromSolcVersion(this.solcConfig.version),
-        outputSelection,
+        outputSelection: this.#dedupeAndSortOutputSelection(outputSelection),
         remappings: this.#remappings.map(formatRemapping),
       },
     };
   }
 
+  #dedupeAndSortOutputSelection(
+    outputSelection: CompilerInput["settings"]["outputSelection"],
+  ): CompilerInput["settings"]["outputSelection"] {
+    const dedupedOutputSelection: CompilerInput["settings"]["outputSelection"] =
+      {};
+
+    for (const [sourceName, contracts] of Object.entries(outputSelection)) {
```

Comment:
I think doing `for (const sourceName of Object.keys(..).sort())` would be safer.

Same with the other for.